File: node/src/http-client.ts
import { TraceData } from "./tracing";
import { JudgeResult } from "./assertions";
import type { SnapshotData } from "./snapshots";

export class VevalHttpClient {
  private readonly endpoint: string;
  private readonly apiKey: string;

  constructor(apiKey: string, endpoint: string) {
    this.apiKey = apiKey;
    this.endpoint = endpoint.replace(/\/$/, "");
  }

  private get headers(): Record<string, string> {
    return {
      "Content-Type": "application/json",
      Authorization: `Bearer ${this.apiKey}`,
    };
  }

  async sendTraceAsync(payload: unknown): Promise<void> {
    try {
      await fetch(`${this.endpoint}/v1/traces`, {
        method: "POST",
        headers: this.headers,
        body: JSON.stringify(payload),
      });
    } catch {
      // swallow silently
    }
  }

  async getTraceAsync(traceId: string): Promise<TraceData | null> {
    try {
      const res = await fetch(`${this.endpoint}/v1/traces/${traceId}`, {
        headers: this.headers,
      });
      if (!res.ok) return null;
      return (await res.json()) as TraceData;
    } catch {
      return null;
    }
  }

  async getScenarioItemsAsync(scenarioName: string): Promise<unknown[]> {
    try {
      const res = await fetch(
        `${this.endpoint}/v1/scenarios/${encodeURIComponent(scenarioName)}/items`,
        { headers: this.headers }
      );
      if (!res.ok) return [];
      const body = (await res.json()) as { items: unknown[] };
      return body.items ?? [];
    } catch {
      return [];
    }
  }

  async postScenarioRunAsync(scenarioName: string, payload: unknown): Promise<void> {
    try {
      await fetch(
        `${this.endpoint}/v1/scenarios/${encodeURIComponent(scenarioName)}/runs`,
        {
          method: "POST",
          headers: this.headers,
          body: JSON.stringify(payload),
        }
      );
    } catch {
      // swallow silently
    }
  }

  /**
   * Saves a named snapshot baseline. Throws on failure — a baseline that silently didn't save would
   * make every later comparison fail, or compare against a stale one.
   */
  async createSnapshotAsync(payload: unknown): Promise<SnapshotData> {
    const res = await fetch(`${this.endpoint}/v1/snapshots`, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify(payload),
    });
    if (!res.ok) {
      throw new Error(`Saving snapshot failed with status ${res.status}: ${await res.text()}`);
    }
    return (await res.json()) as SnapshotData;
  }

  /**
   * Loads the latest baseline with this name. Returns null only when none exists (404); any other
   * failure throws, so a network error can never look like "nothing to compare against".
   */
  async getSnapshotAsync(name: string): Promise<SnapshotData | null> {
    const res = await fetch(`${this.endpoint}/v1/snapshots/${encodeURIComponent(name)}`, {
      headers: this.headers,
    });
    if (res.status === 404) return null;
    if (!res.ok) {
      throw new Error(`Loading snapshot '${name}' failed with status ${res.status}: ${await res.text()}`);
    }
    return (await res.json()) as SnapshotData;
  }

  // Unlike the other calls on this client, judgeAsync deliberately does not swallow
  // failures — a network/API error here must surface as an assertion failure, not a
  // silent pass. Callers are expected to catch.
  async judgeAsync(payload: unknown): Promise<JudgeResult> {
    const res = await fetch(`${this.endpoint}/v1/judge`, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify(payload),
    });
    if (!res.ok) {
      throw new Error(`Judge request failed with status ${res.status}: ${await res.text()}`);
    }
    return (await res.json()) as JudgeResult;
  }
}

File: python/veval/_http_client.py
from __future__ import annotations
import json
from typing import Any, Optional
from urllib.parse import quote
from urllib.request import Request, urlopen
from urllib.error import HTTPError, URLError
from ._tracing import TraceData
from ._snapshots import SnapshotData


class VevalHttpClient:
    def __init__(self, api_key: str, endpoint: str):
        self._api_key = api_key
        self._endpoint = endpoint.rstrip("/")

    def _headers(self) -> dict[str, str]:
        return {
            "Authorization": f"Bearer {self._api_key}",
            "Content-Type": "application/json",
        }

    async def send_trace_async(self, payload: Any) -> None:
        try:
            import asyncio
            await asyncio.get_event_loop().run_in_executor(
                None, self._post, f"{self._endpoint}/v1/traces", payload
            )
        except Exception:
            pass

    async def get_trace_async(self, trace_id: str) -> Optional[TraceData]:
        try:
            import asyncio
            data = await asyncio.get_event_loop().run_in_executor(
                None, self._get, f"{self._endpoint}/v1/traces/{trace_id}"
            )
            return TraceData.from_dict(data) if data else None
        except Exception:
            return None

    async def get_scenario_items_async(self, scenario_name: str) -> list[dict]:
        try:
            import asyncio
            encoded = quote(scenario_name, safe="")
            data = await asyncio.get_event_loop().run_in_executor(
                None, self._get, f"{self._endpoint}/v1/scenarios/{encoded}/items"
            )
            return data.get("items", []) if data else []
        except Exception:
            return []

    async def post_scenario_run_async(self, scenario_name: str, payload: Any) -> None:
        try:
            import asyncio
            encoded = quote(scenario_name, safe="")
            await asyncio.get_event_loop().run_in_executor(
                None, self._post, f"{self._endpoint}/v1/scenarios/{encoded}/runs", payload
            )
        except Exception:
            pass

    async def judge_async(self, payload: Any) -> dict:
        # Unlike the other calls on this client, judge_async deliberately does not swallow
        # failures — a network/API error here must surface as an assertion failure, not a
        # silent pass. Callers are expected to catch.
        import asyncio
        return await asyncio.get_event_loop().run_in_executor(
            None, self._post_json, f"{self._endpoint}/v1/judge", payload
        )

    async def create_snapshot_async(self, payload: Any) -> SnapshotData:
        # Raises on failure — a baseline that silently didn't save would make every later
        # comparison fail, or compare against a stale one.
        import asyncio
        data = await asyncio.get_event_loop().run_in_executor(
            None, self._post_json, f"{self._endpoint}/v1/snapshots", payload
        )
        return SnapshotData.from_dict(data)

    async def get_snapshot_async(self, name: str) -> Optional[SnapshotData]:
        # Returns None only when no baseline exists (404); any other failure raises, so a network
        # error can never look like "nothing to compare against".
        import asyncio
        encoded = quote(name, safe="")
        data = await asyncio.get_event_loop().run_in_executor(
            None, self._get_or_none_on_404, f"{self._endpoint}/v1/snapshots/{encoded}"
        )
        return SnapshotData.from_dict(data) if data is not None else None

    def _get_or_none_on_404(self, url: str) -> Optional[dict]:
        req = Request(url, headers=self._headers(), method="GET")
        try:
            with urlopen(req, timeout=10) as resp:
                return json.loads(resp.read().decode("utf-8"))
        except HTTPError as ex:
            if ex.code == 404:
                return None
            raise

    def _post_json(self, url: str, payload: Any) -> dict:
        body = json.dumps(payload, default=str).encode("utf-8")
        req = Request(url, data=body, headers=self._headers(), method="POST")
        with urlopen(req, timeout=30) as resp:
            return json.loads(resp.read().decode("utf-8"))

    def _post(self, url: str, payload: Any) -> None:
        body = json.dumps(payload, default=str).encode("utf-8")
        req = Request(url, data=body, headers=self._headers(), method="POST")
        try:
            with urlopen(req, timeout=10):
                pass
        except URLError:
            pass

    def _get(self, url: str) -> Optional[dict]:
        req = Request(url, headers=self._headers(), method="GET")
        try:
            with urlopen(req, timeout=10) as resp:
                return json.loads(resp.read().decode("utf-8"))
        except URLError:
            return None
